src/edu/washington/escience/myria/operator/Apply.java

```diff
@@ -153,14 +182,29 @@ protected TupleBatch fetchNextReady() throws DbException, InvocationTargetExcept
             if (!emptyProduct) {
               // Initialize each iterator to its starting index.
               Arrays.fill(iteratorIndexes, 0);
-              // Iterate over each element of the Cartesian product and append to output.
+              // Iterate over each element of the Cartesian product and append to output
+              TupleBuffer countIdx = null;
+              int iteratorIdx = 0;
+              int resultRowIdx = 0;
+              int flatmapid = -1;
               do {
-                for (int iteratorIdx = 0; iteratorIdx < iteratorIndexes.length; ++iteratorIdx) {
-                  int resultRowIdx =
-                      lastCumResultCounts[iteratorIdx] + iteratorIndexes[iteratorIdx];
+                for (iteratorIdx = 0; iteratorIdx < iteratorIndexes.length; ++iteratorIdx) {
+                  resultRowIdx = lastCumResultCounts[iteratorIdx] + iteratorIndexes[iteratorIdx];
+                  if (flatmapid < iteratorIndexes[iteratorIdx]) {
```

Comment:
I don't understand why `flatmapid` is the max of `iteratorIndexes` -- (0,2,0) and (2,2,2) would have the same `flatmapid`?